File: backend/src/svap/stages/stage3_scoring.py
"""
Stage 3: Convergence Scoring & Calibration

Scores each known case against the full taxonomy to produce a convergence
matrix, then analyzes the matrix to identify the threshold where multiple
converging qualities predict large-scale exploitation.

This is the validation stage — if convergence scores don't correlate with
exploitation severity, the taxonomy (Stage 2) needs refinement.

Input:  Cases (Stage 1) + Approved Taxonomy (Stage 2)
Output: Convergence matrix + calibration threshold in `convergence_scores` and `calibration` tables
"""

import json
import logging
from collections import defaultdict

from svap import delta
from svap.bedrock_client import BedrockClient
from svap.rag import ContextAssembler
from svap.storage import SVAPStorage

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """You are scoring a policy against a structural vulnerability taxonomy.
Apply each recognition test precisely. A quality is PRESENT only if the policy clearly
exhibits the structural property described. If ambiguous, mark ABSENT and note the ambiguity
in the evidence field. Do not over-score."""


def _score_case(storage, client, run_id, case, taxonomy_context):
    """Score a single case against the taxonomy and store results."""
    logger.info("Scoring: %s", case['case_name'])
    prompt = client.render_prompt(
        "stage3_score.txt",
        case_name=case["case_name"],
        exploited_policy=case["exploited_policy"],
        scheme_mechanics=case["scheme_mechanics"],
        enabling_condition=case["enabling_condition"],
        taxonomy=taxonomy_context,
    )

    scores = client.invoke_json(prompt, system=SYSTEM_PROMPT, max_tokens=2048)
    score_map = scores.get("scores", scores)

    for quality_id, score_data in score_map.items():
        if isinstance(score_data, dict):
            present = score_data.get("present", False)
            evidence = score_data.get("evidence", "")
        else:
            present = bool(score_data)
            evidence = ""
        storage.insert_convergence_score(run_id, case["case_id"], quality_id, present, evidence)


def _build_calibration_stats(matrix):
    """Build per-case scores, quality frequency, and co-occurrence from the convergence matrix."""
    case_scores = defaultdict(lambda: {"count": 0, "qualities": [], "scale": 0})
    quality_freq = defaultdict(int)

    for row in matrix:
        cid = row["case_id"]
        if case_scores[cid]["scale"] == 0:
            case_scores[cid]["scale"] = row.get("scale_dollars", 0) or 0
            case_scores[cid]["name"] = row["case_name"]
        if row["present"]:
            case_scores[cid]["count"] += 1
            case_scores[cid]["qualities"].append(row["quality_id"])
            quality_freq[row["quality_id"]] += 1

    quality_combos = defaultdict(int)
    for cs in case_scores.values():
        quals = sorted(cs["qualities"])
        for i in range(len(quals)):
            for j in range(i + 1, len(quals)):
                quality_combos[f"{quals[i]}+{quals[j]}"] += 1

    sorted_cases = sorted(case_scores.values(), key=lambda x: x["count"], reverse=True)
    return sorted_cases, dict(quality_freq), dict(quality_combos)


def _run_calibration(client, sorted_cases):
    """Ask the LLM to determine the calibration threshold from case scores."""
    calibration_data = json.dumps(
        [
            {"case": cs["name"], "score": cs["count"], "scale_dollars": cs["scale"]}
            for cs in sorted_cases
        ],
        indent=2,
    )

    cal_prompt = f"""Analyze this convergence score data to determine the calibration threshold.

Each entry shows a known exploitation case, its convergence score (number of vulnerability
qualities present), and the scale in dollars.

{calibration_data}

Determine:
1. THRESHOLD: The convergence score at or above which exploitation tends to be large-scale
   (>$100M+). This is the minimum score that should trigger proactive investigation.
2. CORRELATION_NOTES: Describe the relationship between convergence score and scale.
   Is it linear? Is there a clear step-function? Are there outliers?

Return JSON: {{"threshold": N, "correlation_notes": "..."}}"""

    return client.invoke_json(cal_prompt, max_tokens=1024)


def run(storage: SVAPStorage, client: BedrockClient, run_id: str, config: dict):
    """Execute Stage 3: Score all cases and calibrate."""
    logger.info("Stage 3: Convergence Scoring & Calibration")
    storage.log_stage_start(run_id, 3)

    try:
        stage2_status = storage.get_stage_status(run_id, 2)
        if stage2_status not in ("approved", "completed"):
            raise ValueError(
                f"Stage 2 status is '{stage2_status}'. Taxonomy must be approved before scoring. "
                f"Run: python -m svap.orchestrator approve --stage 2"
            )

        cases = storage.get_cases()
        taxonomy = storage.get_approved_taxonomy()
        ctx = ContextAssembler(storage, config)

        if not cases or not taxonomy:
            raise ValueError("Need both cases (Stage 1) and taxonomy (Stage 2).")

        taxonomy_context = ctx.format_taxonomy_context(taxonomy)

        # ── Delta detection ─────────────────────────────────────────
        tax_fp = delta.taxonomy_fingerprint(taxonomy)
        stored = storage.get_processing_hashes(3)
        cases_to_score = []
        skipped = 0
        for case in cases:
            h = delta.compute_hash(case["enabling_condition"], tax_fp)
            if stored.get(case["case_id"]) == h:
                skipped += 1
            else:
                cases_to_score.append((case, h))

        if not cases_to_score:
            logger.info("All %d cases unchanged. Skipping scoring.", len(cases))
            storage.log_stage_complete(run_id, 3, {"cases_scored": 0, "skipped": skipped})
            logger.info("Stage 3 complete (no changes).")
            return

        # ── Score each case ─────────────────────────────────────────
        logger.info("Scoring %d cases (%d unchanged)...", len(cases_to_score), skipped)
        for case, h in cases_to_score:
            _score_case(storage, client, run_id, case, taxonomy_context)
            storage.record_processing(3, case["case_id"], h, run_id)

        # ── Calibration analysis ────────────────────────────────────
        logger.info("Running calibration analysis...")
        matrix = storage.get_convergence_matrix()
        sorted_cases, quality_freq, quality_combos = _build_calibration_stats(matrix)
        cal_result = _run_calibration(client, sorted_cases)

        threshold = cal_result.get("threshold", 3)
        storage.insert_calibration(
            run_id,
            threshold=threshold,
            notes=cal_result.get("correlation_notes", ""),
            freq=quality_freq,
            combos=quality_combos,
        )

        logger.info("Calibration Results:")
        logger.info("Threshold: %d (policies scoring >=%d are high-priority)", threshold, threshold)
        logger.info("Quality frequency: %s", quality_freq)
        logger.info("Case scores:")
        for cs in sorted_cases:
            marker = "!" if cs["count"] >= threshold else " "
            logger.info("%s %s: score=%d, scale=$%,.0f", marker, cs['name'], cs['count'], cs['scale'])

        storage.log_stage_complete(run_id, 3, {"cases_scored": len(cases), "threshold": threshold})
        logger.info("Stage 3 complete.")

    except Exception as e:
        storage.log_stage_failed(run_id, 3, str(e))
        raise

File: backend/shared/src/stages/stage3_scoring.rs
//! Stage 3: Convergence Scoring & Calibration

use serde_json::json;
use sha2::{Digest, Sha256};
use std::cmp::Reverse;
use std::collections::HashMap;
use tokio_postgres::Client;
use tracing::info;

use crate::bedrock::BedrockClient;
use crate::db;
use crate::rag::ContextAssembler;
use crate::types::{Case, Config, ConvergenceRow, TaxonomyQuality};

const SYSTEM_PROMPT: &str = "You are scoring a policy against a structural vulnerability taxonomy. Apply each recognition test precisely. A quality is PRESENT only if the policy clearly exhibits the structural property described. If ambiguous, mark ABSENT.";

const STAGE3_SCORE_PROMPT: &str = include_str!("../../prompts/stage3_score.txt");
type StageResult<T> = Result<T, Box<dyn std::error::Error + Send + Sync>>;

struct ScoringInputs {
    cases: Vec<Case>,
    taxonomy_context: String,
    tax_fp: String,
}

fn compute_hash(parts: &[&str]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(parts.join("|"));
    format!("{:x}", hasher.finalize())[..12].to_string()
}

fn taxonomy_fingerprint(taxonomy: &[TaxonomyQuality]) -> String {
    let mut ids: Vec<&str> = taxonomy.iter().map(|q| q.quality_id.as_str()).collect();
    ids.sort();
    let mut hasher = Sha256::new();
    hasher.update(ids.join(":"));
    format!("{:x}", hasher.finalize())[..12].to_string()
}

pub async fn run(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    config: &Config,
) -> StageResult<serde_json::Value> {
    info!("Stage 3: Convergence Scoring & Calibration");
    db::log_stage_start(db_client, run_id, 3).await?;

    match run_inner(db_client, bedrock, run_id, config).await {
        Ok(result) => {
            db::log_stage_complete(db_client, run_id, 3, Some(&result)).await?;
            Ok(result)
        }
        Err(e) => {
            db::log_stage_failed(db_client, run_id, 3, &e.to_string()).await?;
            Err(e)
        }
    }
}

async fn run_inner(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    _config: &Config,
) -> StageResult<serde_json::Value> {
    ensure_taxonomy_ready(db_client, run_id).await?;
    let inputs = scoring_inputs(db_client).await?;
    score_and_calibrate(db_client, bedrock, run_id, inputs).await
}

async fn scoring_inputs(db_client: &Client) -> StageResult<ScoringInputs> {
    let cases = db::get_cases(db_client).await?;
    let taxonomy = db::get_approved_taxonomy(db_client).await?;
    if cases.is_empty() || taxonomy.is_empty() {
        return Err("Need both cases and taxonomy.".into());
    }

    Ok(ScoringInputs {
        cases,
        taxonomy_context: ContextAssembler::format_taxonomy_context(&taxonomy),
        tax_fp: taxonomy_fingerprint(&taxonomy),
    })
}

async fn score_and_calibrate(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    inputs: ScoringInputs,
) -> StageResult<serde_json::Value> {
    let stored = db::get_processing_hashes(db_client, 3).await?;
    let (cases_to_score, skipped) = changed_cases(&inputs.cases, &stored, &inputs.tax_fp);

    if cases_to_score.is_empty() {
        info!(
            "All {} cases unchanged. Skipping scoring.",
            inputs.cases.len()
        );
        return Ok(json!({"cases_scored": 0, "skipped": skipped}));
    }

    info!(
        "Scoring {} cases ({} unchanged)...",
        cases_to_score.len(),
        skipped
    );
    score_changed_cases(
        db_client,
        bedrock,
        run_id,
        &inputs.taxonomy_context,
        &cases_to_score,
    )
    .await?;
    let threshold = run_calibration(db_client, bedrock, run_id).await?;

    info!("Stage 3 complete. Threshold: {}", threshold);
    Ok(json!({"cases_scored": cases_to_score.len(), "threshold": threshold}))
}

async fn ensure_taxonomy_ready(db_client: &Client, run_id: &str) -> StageResult<()> {
    let stage2_status = db::get_stage_status(db_client, run_id, 2).await?;
    if matches!(
        stage2_status.as_deref(),
        Some("approved") | Some("completed")
    ) {
        return Ok(());
    }
    Err(format!(
        "Stage 2 status is '{:?}'. Taxonomy must be approved before scoring.",
        stage2_status
    )
    .into())
}

fn changed_cases<'a>(
    cases: &'a [Case],
    stored: &HashMap<String, String>,
    tax_fp: &str,
) -> (Vec<(&'a Case, String)>, usize) {
    let mut changed = Vec::new();
    let mut skipped = 0;

    for case in cases {
        let hash = compute_hash(&[&case.enabling_condition, tax_fp]);
        if stored.get(&case.case_id).map(|s| s.as_str()) == Some(&hash) {
            skipped += 1;
        } else {
            changed.push((case, hash));
        }
    }

    (changed, skipped)
}

async fn score_changed_cases(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    taxonomy_context: &str,
    cases_to_score: &[(&Case, String)],
) -> StageResult<()> {
    for (case, hash) in cases_to_score {
        let scores = score_case(bedrock, case, taxonomy_context).await?;
        insert_case_scores(db_client, run_id, case, &scores).await?;
        db::record_processing(db_client, 3, &case.case_id, hash, run_id).await?;
    }
    Ok(())
}

async fn score_case(
    bedrock: &BedrockClient,
    case: &Case,
    taxonomy_context: &str,
) -> StageResult<serde_json::Value> {
    info!("Scoring: {}", case.case_name);
    let prompt = BedrockClient::render_prompt(
        STAGE3_SCORE_PROMPT,
        &[
            ("case_name", &case.case_name),
            ("exploited_policy", &case.exploited_policy),
            ("scheme_mechanics", &case.scheme_mechanics),
            ("enabling_condition", &case.enabling_condition),
            ("taxonomy", taxonomy_context),
        ],
    );
    bedrock
        .invoke_json(&prompt, SYSTEM_PROMPT, None, Some(2048))
        .await
}

async fn insert_case_scores(
    db_client: &Client,
    run_id: &str,
    case: &Case,
    scores: &serde_json::Value,
) -> StageResult<()> {
    let score_map = scores.get("scores").unwrap_or(scores);
    let Some(obj) = score_map.as_object() else {
        return Ok(());
    };

    for (quality_id, score_data) in obj {
        let (present, evidence) = parse_score(score_data);
        db::insert_convergence_score(
            db_client,
            run_id,
            &case.case_id,
            quality_id,
            present,
            &evidence,
        )
        .await?;
    }
    Ok(())
}

fn parse_score(score_data: &serde_json::Value) -> (bool, String) {
    if let Some(sd) = score_data.as_object() {
        return (
            sd.get("present").and_then(|p| p.as_bool()).unwrap_or(false),
            sd.get("evidence")
                .and_then(|e| e.as_str())
                .unwrap_or("")
                .to_string(),
        );
    }
    (score_data.as_bool().unwrap_or(false), String::new())
}

async fn run_calibration(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
) -> StageResult<i32> {
    info!("Running calibration analysis...");
    let matrix = db::get_convergence_matrix(db_client).await?;
    let (cal_data, quality_freq) = calibration_inputs(&matrix);
    let cal_prompt = format!(
        "Analyze this convergence score data to determine the calibration threshold.\n\n\
         {}\n\n\
         Determine:\n1. THRESHOLD: minimum convergence score for large-scale exploitation\n\
         2. CORRELATION_NOTES: relationship description\n\n\
         Return JSON: {{\"threshold\": N, \"correlation_notes\": \"...\"}}",
        serde_json::to_string_pretty(&cal_data)?
    );
    let cal_result = bedrock
        .invoke_json(&cal_prompt, "", None, Some(1024))
        .await?;
    let threshold = cal_result
        .get("threshold")
        .and_then(|t| t.as_i64())
        .unwrap_or(3) as i32;

    db::insert_calibration(
        db_client,
        run_id,
        threshold,
        cal_result
            .get("correlation_notes")
            .and_then(|n| n.as_str())
            .unwrap_or(""),
        &serde_json::to_value(&quality_freq)?,
        &json!({}),
    )
    .await?;
    Ok(threshold)
}

fn calibration_inputs(matrix: &[ConvergenceRow]) -> (Vec<serde_json::Value>, HashMap<String, i32>) {
    let mut case_scores: HashMap<String, (String, i32, f64)> = HashMap::new();
    let mut quality_freq: HashMap<String, i32> = HashMap::new();

    for row in matrix {
        let entry = case_scores.entry(row.case_id.clone()).or_insert((
            row.case_name.clone(),
            0,
            row.scale_dollars.unwrap_or(0.0),
        ));
        if row.present {
            entry.1 += 1;
            *quality_freq.entry(row.quality_id.clone()).or_insert(0) += 1;
        }
    }

    let mut sorted_cases: Vec<_> = case_scores.values().collect();
    sorted_cases.sort_by_key(|case| Reverse(case.1));
    let cal_data = sorted_cases
        .iter()
        .map(|(name, score, scale)| json!({"case": name, "score": score, "scale_dollars": scale}))
        .collect();
    (cal_data, quality_freq)
}
